### DISCUSSION TITLE
Best practices for using "contextValue"

### DISCUSSION THREAD

**Discussion by illambo** [rebing/graphql-laravel#1126] [Q&A] [answered]
Hi everyone, I'm writing to try to get some information on the best practice for using contextValue.

From [docs](https://webonyx.github.io/graphql-php/class-reference/#graphqlgraphql-methods) I see that it is possible to use this parameter to route a shared object to all resolvers:
> contextValue:
> The context value is provided as an argument to resolver functions after field arguments. It is used to pass shared information useful at any point during executing this query, for example the currently logged in user and connections to databases or other services.
> If the passed object implements the `ScopedContext` interface, its `clone()` method will be called before passing the context down to a field.
> This allows passing information to child fields in the query tree without affecting sibling or parent fields.

Currently this package routes the user object through an [execution middleware](https://github.com/rebing/graphql-laravel/blob/master/src/Support/ExecutionMiddleware/AddAuthUserContextValueMiddleware.php) but one could pass a more structured object/array what he needs (user, tenancy, dataloaders?, ..) even implementing (from what I understand) the [ScopedContext](https://webonyx.github.io/graphql-php/class-reference/#graphqlexecutorscopedcontext) interface if yes wants to keep it “isolated”.
In the case, for example, of a user passed into the context, I therefore wanted to have an opinion on whether it is better at this point to use the information contained within rather than using facade / helper auth in the authorization and miscellaneous control methods of the case which in fact seem to me to bypass the contained value in context (from [docs](https://github.com/rebing/graphql-laravel?tab=readme-ov-file#authorization) it seems not to go through context).

This could allow you to manually force the context (e.g. in tests) in operations like GraphQL::query (which could be a good or bad thing).

What would be the best practices?

Thanks

**Comment by mfn** [comment, marked as answer]
> but one could pass a more structured object/array what he needs (user, tenancy, dataloaders?, ..)

But no one is stopping you from doing so: just return `new YourSuperChargedContext`, you even type hint it in every resolver.

That's what I do and the context holds some requested-shared information, which is quite useful.

Though in recent years I reduced the "shared state" as much, it's like a global var and thus always prone to create hard to predict code.

But anyway, it's possible. Or did I misunderstand?

**Reply by illambo** [reply]
Yes, thank you, I will use the contextValue in this sense, I just wanted to get an expert opinion on the matter. :pray:

**Reply by mfn** [reply]
> expert

Thanks 😼 but that would be quite the exaggeration.

But: I've used this for years this way in prod, big project, works well so far.
